### PR TITLE
PoC, WIP Preconditions as inline functions

### PR DESCRIPTION
**Pull request by antkmsft**
A possible thing that Jeff has proposed in https://github.com/Azure/azure-sdk-for-c/pull/707

### CODE REVIEW

File: sdk/core/core/internal/az_precondition_internal.h [hunk 1]
```diff
@@ -43,23 +43,22 @@
...
-  } while (0)
+  if (!condition)
+  {
+    az_precondition_failed_get_callback()();
+  }
 #endif // AZ_NO_PRECONDITION_CHECKING
+}
 
-#define _az_PRECONDITION_RANGE(low, arg, max) _az_PRECONDITION((low <= arg && arg <= max))
+#define _az_PRECONDITION_RANGE(low, arg, max) _az_precondition((low <= arg && arg <= max))
```
Comment (JeffreyRichter) [inline sdk/core/core/internal/az_precondition_internal.h:58]: Can't this be an inline function too?

Comment (antkmsft) [reply sdk/core/core/internal/az_precondition_internal.h:58]: I'll need to think about it, this is why I left it as a macro :) Maybe it needs to work for signed and unsigned ints of different sizes, and maybe even pointers. I'll take a look at usage. Maybe we'll remove this one at all (convert to conditions you pass to `_az_precondition()`)

File: sdk/core/core/internal/az_precondition_internal.h [hunk 2]
```diff
@@ -43,23 +43,22 @@
 
 az_precondition_failed_fn az_precondition_failed_get_callback();
 
+AZ_INLINE void _az_precondition(bool condition)
```
Comment (ahsonkhan) [inline sdk/core/core/internal/az_precondition_internal.h:46]: I haven't been following the other threads too closely. Can you summarize why we are turning macros into inline functions (maybe in the PR description)? How do we verify this is doing what it wants us to do.

File: sdk/core/core/internal/az_precondition_internal.h (hunk 1 above)
Comment (antkmsft) [reply sdk/core/core/internal/az_precondition_internal.h]: Jeff, I made `_az_precondition_range()` to take int32_t (other option would be to replace all `_az_PRECONDITION_RANGE()` usages with `_az_precondition(low <= arg && arg <= max))`

File: sdk/core/core/internal/az_precondition_internal.h (hunk 2 above)
Comment (antkmsft) [reply sdk/core/core/internal/az_precondition_internal.h:46]: This is a draft and an experiment. I am linking to the PR where Jeff had this idea in comments.
The idea is that with inline functions, the compiler is smart enough to remove the code that has no effect, so we turn macros into inline functions, and they are more readable and better debuggable.

File: sdk/core/core/internal/az_precondition_internal.h (hunk 1 above)
Comment (JeffreyRichter) [reply sdk/core/core/internal/az_precondition_internal.h]: Were all examples you found using int32_t?
No uses with other integer types?

Comment (antkmsft) [reply sdk/core/core/internal/az_precondition_internal.h]: All but one. One is using int16_t.

Comment (JeffreyRichter) [reply sdk/core/core/internal/az_precondition_internal.h]: Great, then this int32_t version should work good enough then.

Comment (antkmsft) [reply sdk/core/core/internal/az_precondition_internal.h]: Yes, I think it might be good enough for most cases. If there's any problem (using it with int64, for example), there should be warnings treated as errors. And for all the non-typical cases (pointers, unsigned, int64s) - they are not common, and can be written as `_az_precondition(min <= x && x <= max)`.
